libraries/adaptive-expressions/src/functionTable.ts

```diff
@@ -0,0 +1,105 @@
+/**
+ * @module adaptive-expressions
+ */
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ExpressionEvaluator } from './expressionEvaluator';
+import { ExpressionFunctions } from './expressionFunctions';
+
+type customFunction = (args: any[]) => any;
+
+/**
+ * FunctionTable is a dictionary which merges BuiltinFunctions.Functions with a CustomDictionary.
+ */
+export class FunctionTable implements Map<string, ExpressionEvaluator> {
+    private readonly customFunctions = new Map<string, ExpressionEvaluator>();
+
+    public keys(): IterableIterator<string> {
+        const keysOfAllFunctions = Array.from(ExpressionFunctions.standardFunctions.keys()).concat(Array.from(this.customFunctions.keys()));
+        return keysOfAllFunctions[Symbol.iterator]();
+    }
+
+    public values(): IterableIterator<ExpressionEvaluator> {
+        const valuesOfAllFunctions = Array.from(ExpressionFunctions.standardFunctions.values()).concat(Array.from(this.customFunctions.values()));
+        return valuesOfAllFunctions[Symbol.iterator]();
+    }
+
+    public get size(): number {
+        return ExpressionFunctions.standardFunctions.size + this.customFunctions.size;
+    }
+
+    public get isReadOnly(): boolean { 
+        return false;
+    }
+
+    public get(key: string): ExpressionEvaluator {
+
+        if(ExpressionFunctions.standardFunctions.get(key)) {
+            return ExpressionFunctions.standardFunctions.get(key);
+        }
+
+        if (this.customFunctions.get(key)) {
+            return this.customFunctions.get(key);
+        }
+
+        return undefined;
+    }
+
+    public set(key: string, value: ExpressionEvaluator): this {
+        if(ExpressionFunctions.standardFunctions.get(key)) {
+            throw Error(`You can't overwrite a built in function.`);
+        }
+
+        this.customFunctions.set(key, value);
+        return this;
+
+    }
+
+    public add(item: {key: string; value: ExpressionEvaluator} | string, value?: ExpressionEvaluator | customFunction): void{
```

Comment:
Maybe use functions overloads is more readable?
https://www.typescriptlang.org/docs/handbook/functions.html#overloads